tests/chainermn_tests/extensions_tests/test_observation_aggregator.py

```diff
@@ -26,64 +27,99 @@ def forward(self, x):
+                                            use_chainer_variable,
+                                            communicate_interval,
+                                            use_gpu=True)
+        else:
+            with pytest.raises(ValueError):
+                run_test_observation_aggregator(communicator, xp,
+                                                use_chainer_variable,
+                                                communicate_interval,
+                                                use_gpu=True)
+
+
+@pytest.mark.parametrize('use_chainer_variable', [True, False])
+@pytest.mark.parametrize('communicate_interval', [1, 2])
+@chainer.testing.attr.gpu
+def test_observation_aggregator_gpu_cupy(use_chainer_variable,
+                                         communicate_interval):
+    communicator = chainermn.create_communicator('pure_nccl')
+
+    run_test_observation_aggregator(communicator, cuda.cupy,
                                     use_chainer_variable,
                                     communicate_interval,
-                                    use_cupy=True)
+                                    use_gpu=True)
 
 
 def run_test_observation_aggregator(comm, xp,
                                     use_chainer_variable,
                                     communicate_interval,
-                                    use_cupy):
+                                    use_gpu):
     model = DummyChain()
-    if use_cupy:
-        model.to_device(cupy.cuda.Device())
-    optimizer = chainermn.create_multi_node_optimizer(
-        chainer.optimizers.Adam(), comm)
-    optimizer.setup(model)
 
-    train = xp.random.rand(10, 1).astype(np.float32)
+    if use_gpu:
+        # Use CuPy's Device class to force call cudaSetDevice()
+        chainer.cuda.get_device_from_id(comm.intra_rank).use()
+
+    device = get_device(comm.intra_rank if use_gpu else None, xp == chainerx)
+    device.use()
+
+    if xp == chainerx:
+        train = xp.array(np.random.rand(10, 1).astype(np.float32))
+    else:
+        train = xp.random.rand(10, 1).astype(np.float32)
+
+    model.to_device(device)
+
     train_iter = chainer.iterators.SerialIterator(train,
                                                   batch_size=1,
                                                   repeat=True,
                                                   shuffle=True)
 
-    updater = chainer.training.StandardUpdater(train_iter, optimizer)
+    optimizer = chainermn.create_multi_node_optimizer(
+        chainer.optimizers.Adam(), comm)
+    optimizer.setup(model)
+
+    updater = chainer.training.StandardUpdater(train_iter, optimizer,
+                                               device=device)
 
     trainer = chainer.training.Trainer(updater, (1, 'epoch'))
 
     @extension.make_extension(
         trigger=(1, 'iteration'), priority=extension.PRIORITY_WRITER)
-    def rank_reporter(trainer):
+    def rank_reporter(trainer_):
```

Comment:
Why did you change `trainer` to `trainer_`?

Reply:
fix closure's parameter name that hid outer variables?
9296d5e14970699aa29619f838918f45c676ffe9

Reply:
Yes, @belldandyxtq is right. if it is `trainer`, it hides the variable in the outer scope.